Only add symbols that are explicit entrypoints when building juliac image

gbaraldi opened the pull request:
This currently just does a linear search over the items. We might want to change this datastructure to a hash table at some point, but given that I don't think libraries  have too too many entrypoints this linear search should be fast enough.

gbaraldi committed:
Only add symbols that are explicit entrypoints

vchuravy commented on src/precompile_utils.c:
```diff
@@ -374,9 +374,8 @@ static void *jl_precompile_trimmed(size_t world)
-    while (1)
-    {
-        mi = (jl_method_instance_t*)arraylist_pop(jl_entrypoint_mis);
+    for (size_t i = 0; i < jl_entrypoint_mis->len ; i++) {
+        mi = (jl_method_instance_t*)jl_entrypoint_mis->items[i];
```
One worry I have here is that there is the potential to break user-code that depends on this behavior.
That code may be rare, but I have seen it in the wild.

JeffBezanson replied:
I think we should stick to default-exclude for everything with trimming; it will keep it easier to use.

topolarity commented on src/staticdata.c:
```diff
@@ -1741,8 +1741,20 @@ static void jl_write_values(jl_serializer_state *s) JL_GC_DISABLED
+                        jl_method_instance_t* mi = (jl_method_instance_t*)jl_entrypoint_mis->items[i];
+                        assert(jl_is_method_instance(mi));
+                        jl_method_t *ccallable = mi->def.method;
+                        if (ccallable == m) {
```
There's a corner case here where the MI != the `@ccallable` signature, which would still lead to a crash I think.

Can we check for that?

gbaraldi replied:
I think we already assert that. 
https://github.com/JuliaLang/julia/blob/fb02e641066fe423c34ee8e4fec24b965662ce85/src/precompile_utils.c#L370-L394

gbaraldi replied:
I can add a comment that mentions that though

topolarity replied:
I don't see where that's asserted there - Can you spell it out for me?

gbaraldi replied:
```julia
static void *jl_precompile_trimmed(size_t world)
{
    // array of MethodInstances and ccallable aliases to include in the output
    jl_array_t *m = jl_alloc_vec_any(0);
    jl_value_t *ccallable = NULL;
    JL_GC_PUSH2(&m, &ccallable);
    jl_method_instance_t *mi;
    for (size_t i = 0; i < jl_entrypoint_mis->len ; i++) {
        mi = (jl_method_instance_t*)jl_entrypoint_mis->items[i];
        if (mi == NULL)
            break;
        assert(jl_is_method_instance(mi));

        jl_array_ptr_1d_push(m, (jl_value_t*)mi);
        ccallable = (jl_value_t *)mi->def.method->ccallable; // If the method passed into jl_entrypoint has ccallable we always compile it. 
        if (ccallable)
            jl_array_ptr_1d_push(m, ccallable);
    }
    jl_cgparams_t params = jl_default_cgparams;
    params.trim = jl_options.trim;
    void *native_code = jl_create_native(m, NULL, &params, 0, /* imaging */ 1, 0,
                                         world, NULL);
    JL_GC_POP();
    return native_code;
}
```

gbaraldi replied:
It means if we are compiling the entrypoint and if the method of that entrypoint has a ccallable defined on it, we will add both the entrypoint MI and the ccallable signature (it's an svec). We can maybe compare ccallables instead of methods.

topolarity replied:
That's fine, but we shouldn't be imprecise about checking the entrypoint just because we happen to enqueue the ccallables in that way right now

gbaraldi committed:
Merge branch 'master' into gb/inline-ccallable

gbaraldi committed:
Merge branch 'master' into gb/inline-ccallable

gbaraldi committed:
Change ccallable handling for entrypoints

gbaraldi committed:
Fix typo

topolarity commented on src/staticdata.c:
```diff
@@ -1788,8 +1788,18 @@ static void jl_write_values(jl_serializer_state *s) JL_GC_DISABLED
-                if (m->ccallable)
-                    arraylist_push(&s->ccallable_list, (void*)reloc_offset);
+                if (m->ccallable) {
+                    int should_push = !jl_options.trim;
+                    for (size_t i = 0; !should_push && i < jl_entrypoint_list->len  ; i++) {
+                        jl_value_t* val = jl_entrypoint_list->items[i];
+                        if ((jl_value_t *)m->ccallable == val) {
+                            should_push = 1;
+                            break;
+                        }
+                    }
+                    if (should_push)
+                        arraylist_push(&s->ccallable_list, (void*)reloc_offset);
```
```suggestion
                    if (jl_options.trim == JL_TRIM_NO) {
                        arraylist_push(&s->ccallable_list, (void*)reloc_offset);
                    } else {
                        for (size_t i = 0; i < jl_entrypoint_list->len; i++) {
                            jl_value_t* val = (jl_value_t*)jl_entrypoint_list->items[i];
                            if ((jl_value_t *)m->ccallable == val) {
                                arraylist_push(&s->ccallable_list, (void*)reloc_offset);
                                break;
                            }
                        }
                    }
```

Just a style suggestion - I find this clearer, despite being a bit redundant

topolarity reviewed (approved):
Looking good! I like this version much better

Thanks @gbaraldi 👍

topolarity commented:
Is it possible to add https://github.com/JuliaLang/julia/issues/56603 as a test?

gbaraldi committed:
Update src/staticdata.c

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

gbaraldi committed:
Add test for ccallable function called by ccallable function.

inkydragon commented:
Test failed seems related.

```
gcc -march=native -mtune=native -m64 -c -o init.o /cache/build/tester-amdci4-8/julialang/julia-master/test/trimming/init.c   -std=gnu11 -I'/cache/build/tester-amdci4-8/julialang/julia-master/usr/include/julia' -fPIC
make: Leaving directory '/cache/build/tester-amdci4-8/julialang/julia-master/test/trimming'
make: Entering directory '/cache/build/tester-amdci4-8/julialang/julia-master/test/trimming'
/cache/build/tester-amdci4-8/julialang/julia-master/usr/bin/julia -t 1 -J /cache/build/tester-amdci4-8/julialang/julia-master/usr/bin/../lib/julia/sys.so --startup-file=no --history-file=no --output-o hello.o --output-incremental=no --strip-ir --strip-metadata --experimental --trim /cache/build/tester-amdci4-8/julialang/julia-master/usr/bin/../share/julia/juliac-buildscript.jl /cache/build/tester-amdci4-8/julialang/julia-master/test/trimming/hello.jl --output-exe true
warning: no code provided for function main() from main()
[171] signal 6 (-6): Aborted
in expression starting at none:0
gsignal at /lib/x86_64-linux-gnu/libc.so.6 (unknown line)
```
https://buildkite.com/julialang/julia-master/builds/43995#0194929f-3352-4899-a516-022a0ca44b27/790-809

JeffBezanson committed:
fix for ccallables with trimming

JeffBezanson commented:
I pushed a version that should work. However there still seems to be a bug with const_return ccallables; including `foo` as an entrypoint in this test crashes.

JeffBezanson committed:
include CodeInstances for ccallables with const_api in inference output

This is only needed because `jl_reinit_ccallable` looks at method
metadata that trimming will normally remove for const_api code.

JeffBezanson committed:
include CodeInstances with const_api in inference output when trimming

JeffBezanson commented:
OK I fixed that and added a test. But, we could ignore this whole situation entirely by skipping `jl_reinit_ccallable` for trimmed binaries. If you deploy without libjulia-codegen, it won't do anything anyway!

topolarity commented on Compiler/src/typeinfer.jl:
```diff
@@ -1292,7 +1294,7 @@ function typeinf_ext_toplevel(methods::Vector{Any}, worlds::Vector{UInt}, trim::
             def = mi.def
             if use_const_api(callee)
-                src = codeinfo_for_const(interp, mi, code.rettype_const)
+                src = codeinfo_for_const(interp, mi, callee.rettype_const)
```
Were we not hitting this code path before?

That's... kind of frightening

JeffBezanson replied:
Yep looks like it!

JeffBezanson committed:
Merge branch 'gb/inline-ccallable' of github.com:JuliaLang/julia into gb/inline-ccallable
